File: kotlin/bc-envelope/src/test/kotlin/com/blockchaincommons/bcenvelope/MultiPermitTest.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.*
import com.blockchaincommons.dcbor.CborDate
import com.blockchaincommons.knownvalues.DATE
import kotlin.test.Test
import kotlin.test.assertEquals

class MultiPermitTest {

    @Test
    fun testMultiPermit() {
        registerTags()

        // Alice composes a poem.
        val poemText =
            "At midnight, the clocks sang lullabies to the wandering teacups."

        // Alice creates a new envelope and assigns the text as the subject.
        val originalEnvelope = Envelope.from(poemText)
            .addType("poem")
            .addAssertion("title", "A Song of Ice Cream")
            .addAssertion("author", "Plonkus the Iridescent")
            .addAssertion(DATE, CborDate.fromYmd(2025, 5, 15))

        // Alice signs the envelope with her private key.
        val (alicePrivateKeys, alicePublicKeys) = keypair()
        val signedEnvelope = originalEnvelope.sign(alicePrivateKeys)

        val expectedSignedFormat = """
            {
                "At midnight, the clocks sang lullabies to the wandering teacups." [
                    'isA': "poem"
                    "author": "Plonkus the Iridescent"
                    "title": "A Song of Ice Cream"
                    'date': 2025-05-15
                ]
            } [
                'signed': Signature
            ]
        """.trimIndent()
        assertEquals(expectedSignedFormat, signedEnvelope.format())

        // Alice picks a random content key and encrypts the signed envelope.
        val contentKey = SymmetricKey.create()
        val encryptedEnvelope = signedEnvelope.encrypt(contentKey)
        assertEquals("ENCRYPTED", encryptedEnvelope.format())

        // Alice adds the first permit using her password.
        val password = "unicorns_dance_on_mars_while_eating_pizza"
        val lockedEnvelope = encryptedEnvelope
            .addSecret(KeyDerivationMethod.Argon2id, password, contentKey)

        val expectedLockedFormat = """
            ENCRYPTED [
                'hasSecret': EncryptedKey(Argon2id)
            ]
        """.trimIndent()
        assertEquals(expectedLockedFormat, lockedEnvelope.format())

        // Alice adds recipient permits for herself and Bob.
        val (bobPrivateKeys, bobPublicKeys) = keypair()
        val lockedWithRecipients = lockedEnvelope
            .addRecipient(alicePublicKeys, contentKey)
            .addRecipient(bobPublicKeys, contentKey)

        val expectedRecipientsFormat = """
            ENCRYPTED [
                'hasRecipient': SealedMessage
                'hasRecipient': SealedMessage
                'hasSecret': EncryptedKey(Argon2id)
            ]
        """.trimIndent()
        assertEquals(expectedRecipientsFormat, lockedWithRecipients.format())

        // Alice creates a 2-of-3 SSKR split.
        val sskrGroup = SSKRGroupSpec(2, 3)
        val spec = SSKRSpec(1, listOf(sskrGroup))
        val shardedEnvelopes =
            lockedWithRecipients.sskrSplitFlattened(spec, contentKey)

        val expectedShardedFormat = """
            ENCRYPTED [
                'hasRecipient': SealedMessage
                'hasRecipient': SealedMessage
                'hasSecret': EncryptedKey(Argon2id)
                'sskrShare': SSKRShare
            ]
        """.trimIndent()
        assertEquals(expectedShardedFormat, shardedEnvelopes[0].format())

        // Five ways to unlock:

        // 1. Using the content key.
        val receivedEnvelope = shardedEnvelopes[0]
        val unlockedByKey = receivedEnvelope.decrypt(contentKey)
        assertEquals(signedEnvelope, unlockedByKey)

        // 2. Using the password.
        val unlockedByPassword = receivedEnvelope.unlock(password)
        assertEquals(signedEnvelope, unlockedByPassword)

        // 3. Using Alice's private key.
        val unlockedByAlice = receivedEnvelope.decryptToRecipient(alicePrivateKeys)
        assertEquals(signedEnvelope, unlockedByAlice)

        // 4. Using Bob's private key.
        val unlockedByBob = receivedEnvelope.decryptToRecipient(bobPrivateKeys)
        assertEquals(signedEnvelope, unlockedByBob)

        // 5. Using any two of the three SSKR shares.
        val unlockedBySskr =
            Envelope.sskrJoin(listOf(shardedEnvelopes[0], shardedEnvelopes[2]))
                .unwrap()
        assertEquals(signedEnvelope, unlockedBySskr)

        unlockedBySskr.verify(alicePublicKeys)
    }
}

File: go/bcenvelope/multi_permit_test.go
package bcenvelope

import (
	"testing"

	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
	knownvalues "github.com/nickel-blockchaincommons/knownvalues-go"
	sskr "github.com/nickel-blockchaincommons/sskr-go"
)

func TestMultiPermit(t *testing.T) {
	bccomponents.RegisterTags()

	//
	// Alice composes a poem.
	//
	poemText := "At midnight, the clocks sang lullabies to the wandering teacups."

	//
	// Alice creates a new envelope and assigns the text as the envelope's
	// subject.
	//
	date := dcbor.DateFromYMD(2025, 5, 15)
	originalEnvelope := NewEnvelope(poemText).
		AddType("poem").
		AddAssertion("title", "A Song of Ice Cream").
		AddAssertion("author", "Plonkus the Iridescent").
		AddAssertion(knownvalues.Date, date)

	//
	// Alice signs the envelope with her private key.
	//
	alicePrivKeys, alicePubKeys, err := bccomponents.Keypair()
	if err != nil {
		t.Fatalf("keypair failed: %v", err)
	}
	signedEnvelope := originalEnvelope.Sign(alicePrivKeys)

	assertActualExpected(t, signedEnvelope.Format(), `{
    "At midnight, the clocks sang lullabies to the wandering teacups." [
        'isA': "poem"
        "author": "Plonkus the Iridescent"
        "title": "A Song of Ice Cream"
        'date': 2025-05-15
    ]
} [
    'signed': Signature
]`)

	//
	// Alice picks a random symmetric "content key" and uses it to encrypt the
	// signed envelope.
	//
	contentKey := bccomponents.NewSymmetricKey()
	encryptedEnvelope := signedEnvelope.Encrypt(contentKey)

	assertActualExpected(t, encryptedEnvelope.Format(), `ENCRYPTED`)

	//
	// Alice wants to be able to recover the envelope later using a password.
	//
	password := []byte("unicorns_dance_on_mars_while_eating_pizza")
	lockedEnvelope, err := encryptedEnvelope.AddSecret(
		bccomponents.KDMethodArgon2id,
		password,
		contentKey,
	)
	if err != nil {
		t.Fatalf("add_secret failed: %v", err)
	}

	assertActualExpected(t, lockedEnvelope.Format(), `ENCRYPTED [
    'hasSecret': EncryptedKey(Argon2id)
]`)

	//
	// Next, Alice wants to be able to unlock her envelope using her private
	// key, and she also wants Bob to be able to unlock it using his private
	// key.
	//
	bobPrivKeys, bobPubKeys, err := bccomponents.Keypair()
	if err != nil {
		t.Fatalf("bob keypair failed: %v", err)
	}
	lockedEnvelope = lockedEnvelope.
		AddRecipient(alicePubKeys.EncapsulationPublicKey(), contentKey).
		AddRecipient(bobPubKeys.EncapsulationPublicKey(), contentKey)

	assertActualExpected(t, lockedEnvelope.Format(), `ENCRYPTED [
    'hasRecipient': SealedMessage
    'hasRecipient': SealedMessage
    'hasSecret': EncryptedKey(Argon2id)
]`)

	//
	// Alice creates a 2-of-3 SSKR group and "shards" the envelope.
	//
	sskrGroup, err := sskr.NewGroupSpec(2, 3)
	if err != nil {
		t.Fatalf("sskr group spec failed: %v", err)
	}
	sskrSpec, err := sskr.NewSpec(1, []sskr.GroupSpec{sskrGroup})
	if err != nil {
		t.Fatalf("sskr spec failed: %v", err)
	}
	shardedEnvelopes, err := lockedEnvelope.SSKRSplitFlattened(&sskrSpec, contentKey)
	if err != nil {
		t.Fatalf("sskr split failed: %v", err)
	}

	assertActualExpected(t, shardedEnvelopes[0].Format(), `ENCRYPTED [
    'hasRecipient': SealedMessage
    'hasRecipient': SealedMessage
    'hasSecret': EncryptedKey(Argon2id)
    'sskrShare': SSKRShare
]`)

	//
	// Five different ways to unlock:
	//

	// 1. Using the content key.
	receivedEnvelope := shardedEnvelopes[0]
	unlockedEnvelope, err := receivedEnvelope.Decrypt(contentKey)
	if err != nil {
		t.Fatalf("decrypt with content key failed: %v", err)
	}
	if !unlockedEnvelope.IsEquivalentTo(signedEnvelope) {
		t.Fatal("content key unlocked envelope should equal signed envelope")
	}

	// 2. Using the password.
	unlockedEnvelope, err = receivedEnvelope.Unlock(password)
	if err != nil {
		t.Fatalf("unlock with password failed: %v", err)
	}
	if !unlockedEnvelope.IsEquivalentTo(signedEnvelope) {
		t.Fatal("password unlocked envelope should equal signed envelope")
	}

	// 3. Using Alice's private key.
	unlockedEnvelope, err = receivedEnvelope.DecryptToRecipient(alicePrivKeys.EncapsulationPrivateKey())
	if err != nil {
		t.Fatalf("decrypt to Alice recipient failed: %v", err)
	}
	if !unlockedEnvelope.IsEquivalentTo(signedEnvelope) {
		t.Fatal("Alice recipient unlocked envelope should equal signed envelope")
	}

	// 4. Using Bob's private key.
	unlockedEnvelope, err = receivedEnvelope.DecryptToRecipient(bobPrivKeys.EncapsulationPrivateKey())
	if err != nil {
		t.Fatalf("decrypt to Bob recipient failed: %v", err)
	}
	if !unlockedEnvelope.IsEquivalentTo(signedEnvelope) {
		t.Fatal("Bob recipient unlocked envelope should equal signed envelope")
	}

	// 5. Using any two of the three SSKR shares.
	sskrJoined, err := SSKRJoin([]*Envelope{shardedEnvelopes[0], shardedEnvelopes[2]})
	if err != nil {
		t.Fatalf("sskr join failed: %v", err)
	}
	unlockedEnvelope, err = sskrJoined.TryUnwrap()
	if err != nil {
		t.Fatalf("unwrap failed: %v", err)
	}
	if !unlockedEnvelope.IsEquivalentTo(signedEnvelope) {
		t.Fatal("SSKR unlocked envelope should equal signed envelope")
	}

	// Verify Alice's signature
	_, err = unlockedEnvelope.Verify(alicePubKeys)
	if err != nil {
		t.Fatalf("signature verification failed: %v", err)
	}
}
